Use the ConversionManager from Castle.Windsor

dvdwouwe opened the pull request:
Hi,

I think that the build-in conversion manager of Castle.Windsor is better then using the current code. You can easily enhance the build-in conversion manager of Castle.Windsor. 
Out of the box more and complex types are supported. 

I also improved the EnvironmentProvider to check 3 levels:

1. Process
2. User
3. Machine

To add new types, for example the Version class:

```c#
// Register all registered ITypeConverter services to the default Conversion Manager
ITypeConverter[] typeConverters = container.ResolveAll<ITypeConverter>();
IConversionManager conversionManager = container.Kernel.GetConversionManager();
foreach (ITypeConverter converter in typeConverters)
{
    conversionManager.Add(converter);
    container.Release(converter);
}

// Base class for converters
public abstract class AbstractTypeConverter : ITypeConverter
  {
    public ITypeConverterContext Context { get; set; }

    public abstract bool CanHandleType(Type type);

    public abstract object PerformConversion(string value, Type targetType);

    public abstract object PerformConversion(IConfiguration configuration, Type targetType);

    /// <summary>
    ///   Returns true if this instance of <c>ITypeConverter</c>
    ///   is able to handle the specified type with the specified
    ///   configuration
    /// </summary>
    /// <param name="type"></param>
    /// <param name="configuration"></param>
    /// <returns></returns>
    /// <remarks>
    ///   The default behavior is to just pass it to the normal CanHadnleType
    ///   peeking into the configuration is used for some advanced functionality
    /// </remarks>
    public virtual bool CanHandleType(Type type, IConfiguration configuration)
    {
      return this.CanHandleType(type);
    }

    public TTarget PerformConversion<TTarget>(string value)
    {
      return (TTarget) this.PerformConversion(value, typeof (TTarget));
    }

    public TTarget PerformConversion<TTarget>(IConfiguration configuration)
    {
      return (TTarget) this.PerformConversion(configuration, typeof (TTarget));
    }
  }


// Version Type converter
public class VersionTypeConverter : AbstractTypeConverter
    {
        public override bool CanHandleType(Type type)
        {
            return type == typeof(Version);
        }

        public override object PerformConversion(IConfiguration configuration, Type targetType)
        {
            return PerformConversion(configuration.Value, targetType);
        }

        public override object PerformConversion(string value, Type targetType)
        {
            return Version.Parse(value);
        }
    }
```

adamconnelly commented:
Hey Danny,

Thanks a lot for taking the time to put together this PR. From a quick glance it looks good, but I just want to take the time to pull it down and have a look in a bit more detail as soon as I've got a chance.

In the meantime, would you mind adding a couple of unit tests for the additional environment variable scenarios you added?

dvdwouwe commented:
After some thinking/reading about the EnvironmentSettingsProvider it is maybe the best to revert the lines
```c#
            return
                Environment.GetEnvironmentVariable(key, EnvironmentVariableTarget.Process)
                ?? Environment.GetEnvironmentVariable(key, EnvironmentVariableTarget.User)
                ?? Environment.GetEnvironmentVariable(key, EnvironmentVariableTarget.Machine);
```

back to
```c#
            return Environment.GetEnvironmentVariable(key);
```
for the following reasons:

- Only Process is supported by Linux / Ios, thinking about to support .netstandard, Castle is already ready for this.

- Additional security is needed when accessing Machine / User scopes, (new EnvironmentPermission(PermissionState.Unrestricted)).Demand()

We changed this provider because of this:
https://stackoverflow.com/questions/12754884/environment-getenvironmentvariable-wont-find-variable-value
It solved the restart issue, but we now have additional hidden access to the registry.

Should I revert the EnvironmentSettingsProvider or leave it?

PS
If you want I can change (another PR) to have support for .netstandard, so your facility can also being used on macOS and Linux environments.

adamconnelly commented:
I think it makes sense to revert the change to `EnvironmentSettingsProvider`. I think it's fairly normal behaviour for a process to only have the environment variables that existed when it started, and I wouldn't want this change to solve the specific case of IIS, but break the simple scenario where you just care about getting the current process environment variables.

But maybe you could create a copy of the provider in your own code base, and then initialise the facility like this:

```csharp
container.AddFacility<AppConfigFacility>(f => f.AddSettingsProvider<CustomEnvironmentSettingsProvider>());
```

As for .netstandard - if you want to create a PR for that I have no problem with that. I had kind of assumed that I would just stop using the facility when moving to .NET core since it has equivalent functionality out of the box via the Configuration API, but I would be more than happy if you wanted to add support for the facility since I don't have much experience with .netstandard yet.

Cheers for looking into how `Environment. GetEnvironmentVariable()` works.

dvdwouwe commented:
Hi,

I reverted the code of the EnvironmentSettingsProvider and added two more examples what the library can do.

Quote:
> I had kind of assumed that I would just stop using the facility when moving to .NET core since it has equivalent functionality out of the box via the Configuration API

Can you hint where I kind find more information about this Configuration API.

By the way, your library is also capable of doing fallbacks, and that is powerfull:
- first check app config
- if previous attempt failed, then check environment vars
- if previous attempt failed, then check an webApi
- ....

I can create a PR for supporting more frameworks, aka .netstandard, only I don't have any experience with the current build platform of Github, nl. Travis, can you assist with that?

adamconnelly commented:
@dvdwouwe I've pulled down the code and tested it now - it's working fine, so I'm happy to merge it. I made some tweaks because I wanted to try to keep the separation between the unit and integration tests, and wanted to avoid passing the full `IKernel` object around. I've put the changes up at https://github.com/adamconnelly/WindsorAppConfigFacility/pull/5. Can you take a look, and if you're ok with them I'll merge them in and publish a new version of the facility to NuGet.

Thanks again for the contribution.

adamconnelly commented:
That's it merged, built and published to NuGet. The package is available at https://www.nuget.org/packages/WindsorAppConfigFacility/1.1.0.

Thanks again for contributing!

Closing since the changes were merged as part of #5.
